src/main/webapp/app/exam/monitoring/charts/monitoring-chart.ts

```diff
@@ -0,0 +1,128 @@
+import { Color } from '@swimlane/ngx-charts';
+import dayjs from 'dayjs/esm';
+
+// Collection of colors used for the exam monitoring
+const colors = [GraphColors.LIGHT_BLUE, GraphColors.LIGHT_GREY, GraphColors.BLUE, GraphColors.GREY, GraphColors.DARK_BLUE];
+
+/**
+ * Returns a color based on the index (modulo).
+ * @param index
+ * @return selected color based on index
+ */
+export function getColor(index: number) {
+    return colors[index % colors.length];
+}
+
+/**
+ * Groups actions by ceiled timestamp or timestamp.
+ * @param examActions array of actions
+ * @return grouped actions
+ */
+export function groupActionsByTimestamp(examActions: ExamAction[]) {
+    return groupBy(examActions, (action: ExamAction) => {
+        return action.ceiledTimestamp ?? action.timestamp;
+    });
+}
+
+/**
+ * Groups actions by type.
+ * @param examActions array of actions
+ * @return grouped actions
+ */
+export function groupActionsByType(examActions: ExamAction[]) {
+    return groupBy(examActions, (action: ExamAction) => {
+        return action.type;
+    });
+}
+
+/**
+ * Groups actions by activity id.
+ * @param examActions array of actions
+ * @return grouped actions
+ */
+export function groupActionsByActivityId(examActions: ExamAction[]) {
+    return groupBy(examActions, (action: ExamAction) => {
+        return action.examActivityId;
+    });
+}
+
+/**
+ * Get the last action grouped by activity id.
+ * @param examActions array of actions
+ * @return filtered and grouped actions
+ */
+export function getLastActionGroupedByActivityId(examActions: ExamAction[]) {
+    const activityActionMap: Map<number, ExamAction> = new Map();
+    for (const [key, actions] of Object.entries(groupActionsByActivityId(examActions))) {
+        activityActionMap.set(Number(key), (actions.sort((a, b) => (a.timestamp?.isAfter(b.timestamp) ? 1 : -1)) as ExamAction[]).last()!);
+    }
+    return activityActionMap;
+}
+
+/**
+ * Get actions filtered for SavedExerciseAction and grouped by activity id.
+ * @param examActions array of actions
+ * @return filtered and grouped actions
+ */
+export function getSavedExerciseActionsGroupedByActivityId(examActions: ExamAction[]) {
+    return groupActionsByActivityId(examActions.filter((action) => action.type === ExamActionType.SAVED_EXERCISE));
+}
+
+/**
+ * Get actions filtered for SwitchedExerciseAction and grouped by activity id.
+ * @param examActions array of actions
+ * @return filtered and grouped actions
+ */
+export function getSwitchedExerciseActionsGroupedByActivityId(examActions: ExamAction[]) {
+    return groupActionsByActivityId(examActions.filter((action) => action.type === ExamActionType.SWITCHED_EXERCISE));
+}
+
+/**
+ * Returns the current amount of students per exercise.
+ * @param examActions array of actions
+ * @return amount of students per exercise as map
+ */
+export function getCurrentAmountOfStudentsPerExercises(examActions: ExamAction[]): Map<number, number> {
```

Comment:
Why did you only specify the return type here?
Especially for developers like me that are not as profund in Typescript, I think the return type would help in all methods